"""
Extra steps with the host reads 
    - map to the the reference mitogenome
"""
from glob import glob

rule host_mito_mapping:
    input:
        mr1 = os.path.join(dir_hostcleaned,"{sample}_R1.mapped.fastq.gz"),
        mr2 = os.path.join(dir_hostcleaned,"{sample}_R2.mapped.fastq.gz"),
        host= config['args']['host_seq']
    output:
        mr1_mt = os.path.join(dir_hostcleaned, "mitogenome", "{sample}_mt_R1.mapped.fastq.gz"),
        mr2_mt = os.path.join(dir_hostcleaned, "mitogenome", "{sample}_mt_R2.mapped.fastq.gz"),
        mapped_bam=os.path.join(dir_hostcleaned, "mitogenome", "{sample}_mapped.bam"),
    params:
        stats=os.path.join(dir_hostcleaned, "mitogenome", "{sample}_bamstats.txt")
    conda:
        os.path.join(dir_env, "minimap2.yaml")
    resources:
        mem_mb =config['resources']['smalljob']['mem_mb'],
        runtime = config['resources']['smalljob']['runtime']
    threads: 
        config['resources']['smalljob']['threads']
    shell:
        """
        set -euo pipefail
        if [ -f {output.mr1_mt} ] && [ -f {output.mr2_mt} ]; then
            echo "Host mitogenome mapping already done. Skipping..."
            exit 0
        else
            minimap2 -ax sr -t {threads} {input.host} {input.mr1} {input.mr2} \
                | samtools view -b -F 4 -@ {threads} -o {output.mapped_bam} -
            samtools flagstat {output.mapped_bam} > {params.stats}

            samtools fastq -@ {threads} -0 /dev/null -s /dev/null -n \
                -1 >(gzip -c  > {output.mr1_mt}) \
                -2 >(gzip -c  > {output.mr2_mt}) \
                {output.mapped_bam}
            

            touch {output.mr1_mt}
            touch {output.mr2_mt}
        
        fi
        """

rule host_mito_snps:
    input:
        bam = os.path.join(dir_hostcleaned, "mitogenome", "{sample}_mapped.bam"),
        host= config['args']['host_seq']
    output:
        vcf = os.path.join(dir_hostcleaned, "mitogenome", "{sample}_mitogenome_snps.vcf.gz"),
        filterred_vcf = os.path.join(dir_hostcleaned, "mitogenome", "{sample}_mitogenome_snps.filtered.vcf.gz")
    params:
        sort_bam=os.path.join(dir_hostcleaned, "mitogenome", "{sample}_mapped.sorted.bam"),
        prefix = os.path.join(dir_hostcleaned, "mitogenome", "{sample}_mitogenome"),
        stats=os.path.join(dir_hostcleaned, "mitogenome", "{sample}_mitogenome_snps.stats.txt"),
        depth=os.path.join(dir_hostcleaned, "mitogenome", "{sample}_mitogenome_snps_depth.txt")
    conda:
        os.path.join(dir_env, "minimap2.yaml")
    resources:
        mem_mb =config['resources']['smalljob']['mem_mb'],
        runtime = config['resources']['smalljob']['runtime']
    threads: 
        config['resources']['smalljob']['threads']
    shell:
        """
        set -euo pipefail
        if [ -f {output.vcf} ]; then
            echo "Output vcf file found, so this run looks liks its run. Skipping..."
            exit 0
        else

            #prep the bam
            samtools sort -o {params.sort_bam} {input.bam}
            samtools index {params.sort_bam}

            #call variants
            bcftools mpileup -f {input.host} -Q 20 -q 20 {params.sort_bam} | \
                bcftools call --ploidy 1 -mv -Ov -o {output.vcf}
            bcftools index {output.vcf}

            #filter the SNPs, QUAL>30 means 0.1% error rate, DP>10 means at least 10 reads support
            bcftools filter -i 'QUAL>30 && DP>10' {output.vcf} -Oz -o {output.filterred_vcf}
            bcftools index {output.filterred_vcf}

            #getting the stats
            bcftools stats {output.filterred_vcf} > {params.stats}
            samtools depth {params.sort_bam} > {params.depth}
        fi
        """

rule normalise_vcfs:
    input:
        filterred_vcf = os.path.join(dir_hostcleaned, "mitogenome", "{sample}_mitogenome_snps.filtered.vcf.gz")
    output:
        norm_vcf = os.path.join(dir_hostcleaned, "mitogenome", "{sample}_mitogenome_snps.filtered.norm.vcf.gz")
    params:
        host= config['args']['host_seq'],
        temp=os.path.join(dir_hostcleaned, "mitogenome", "{sample}_mitogenome_snps.filtered.temp.vcf.gz"), 
    conda:
        os.path.join(dir_env, "minimap2.yaml")
    shell:
        """
        set -euo pipefail
        if [ -f {output.norm_vcf} ]; then
            echo "Normalized VCF already exists. Skipping..."
            exit 0
        else
            bcftools norm -f {params.host} -m -any {input.filterred_vcf} -Oz -o {params.temp} 
            bcftools index {params.temp}

            #then keep only SNPs
            bcftools view -v snps {params.temp} -Oz -o {output.norm_vcf}
            bcftools index {output.norm_vcf}
        fi
        """

rule generate_allele_frequency:
    input:
        norm_vcf = os.path.join(dir_hostcleaned, "mitogenome", "{sample}_mitogenome_snps.filtered.norm.vcf.gz")
    output:
        af_table = os.path.join(dir_hostcleaned, "mitogenome", "{sample}_mitogenome_snps.allele_frequency.txt")
    params:
        prefix = "{sample}",
    conda:
        os.path.join(dir_env, "minimap2.yaml")
    shell:
        """
        set -euo pipefail
        if [ -f {output.af_table} ]; then
            echo "Allele frequency table already exists. Skipping..."
            exit 0
        else
            bcftools query -f '%POS\t[%AD]\n' "$vcf" |   awk -v s={params.prefix} '{
                split($2,a,",");
                if (length(a) == 2) {
                total = a[1] + a[2];
                if (total > 0)
                    print s "\t" $1 "\t" a[2]/total
                }
            } > {output.af_table}
        fi
        """

rule merge_vcf:
    input:
        expand(os.path.join(dir_hostcleaned, "mitogenome", "{sample}_mitogenome_snps.filtered.norm.vcf.gz"), sample=sample_names)
    output:
        merged_vcf = os.path.join(dir_hostcleaned, "mitogenome", "merged_mitogenome_snps.filtered.norm.vcf.gz")
    params:
        temp=os.path.join(dir_hostcleaned, "mitogenome", "merged_mitogenome_snps.temp.vcf.gz"),
        folder=os.path.join(dir_hostcleaned, "mitogenome")
    conda:
        os.path.join(dir_env, "minimap2.yaml")
    shell:
        """
        set -euo pipefail
        if [ -f {output.merged_vcf} ]; then
            echo "Merged VCF already exists. Skipping..."
            exit 0
        else
            bcftools merge -Oz -o {params.temp} {params.folder}/*_mitogenome_snps.filtered.norm.vcf.gz

            mv {params.temp} {output.merged_vcf}
            bcftools index {output.merged_vcf}
        fi
        """

rule snp_alignment:
    input:
        merged_vcf = os.path.join(dir_hostcleaned, "mitogenome", "merged_mitogenome_snps.filtered.norm.vcf.gz"),
        host= config['args']['host_seq']
    output:
        consensus_fasta = os.path.join(dir_reports, "mitogenome", "{sample}_consensus.fasta")
    params:
        sample = "{sample}",
    conda:
        os.path.join(dir_env, "minimap2.yaml")
    params:
        sample="{sample}",
        folder=os.path.join(dir_hostcleaned, "mitogenome")
    shell:
        """
        set -euo pipefail
        if [ -f {output.consensus_fasta} ]; then
            echo "SNP alignment fasta already exists. Skipping..."
            exit 0
        else
            SAMPLE_FULL=$(bcftools query -l {input.merged_vcf} | grep "{params.sample}")
            bcftools consensus -s "$SAMPLE_FULL" -f {input.host} {input.merged_vcf} > {output.consensus_fasta}

            # Add sample name to fasta header
            sample="{params.sample}"
            sed -i "s/>/>${{sample}}_/g" {output.consensus_fasta}
        fi
        """

